### PR TITLE
Display documents from packages in the dependencies tree

### LINKED ISSUE #9621: Dependencies tree unit tests
From https://github.com/NuGet/NuGet.Client/pull/3392

Tracking adding more tests to the code in NuGet.Client for populating transitive dependencies in Solution Explorer.

### LINKED ISSUE #9622: Dependencies tree Apex tests
From https://github.com/NuGet/NuGet.Client/pull/3392

Investigate incorporating Dependencies tree Apex tests for Solution Explorer.

The .NET Project System has some tests that may help here.

https://github.com/dotnet/project-system/blob/e3944ffa90462ea031e4754303ff554fe8f794f9/tests/Microsoft.VisualStudio.ProjectSystem.IntegrationTests/DependencyNodeIntegrationTests.cs

### PR DESCRIPTION
**Pull request by drewnoakes**
## Bug

Fixes: https://github.com/NuGet/Home/issues/9621
Fixes: https://github.com/NuGet/Home/issues/9622
Fixes: https://developercommunity.visualstudio.com/t/open-package-readmeme-from-Dependencies/744260

Regression? Last working version: N/A

## Description

Packages may contain files such as README files, license text, etc. This change adds a "Documents" node beneath packages in the dependencies tree of Solution Explorer, when such files exist, allowing them to be opened and viewed in the editor.

It's also possible to search in Solution Explorer for these items, so long as the "Search within external items" option is checked.

image

image

image

image

## PR Checklist

- [x] PR has a meaningful title
- [x] PR has a linked issue.
- [x] Described changes

- **Tests**
  - [ ] Automated tests added
  - **OR**
  
  - [x] Test exception
    - These files are mostly declarative relationships between entities. There's no current test infrastructure for these.
  - **OR**
  - [ ] N/A 

- **Documentation**
  
  - [ ] Documentation PR or issue filled
  - **OR**
  - [x] N/A

### CODE REVIEW

**Comment by nkolev92** [comment]
NuGet packages have the concepts of license and readme from within the package. An alternative could be to explicitly lookup those. 

https://learn.microsoft.com/en-us/nuget/reference/nuspec#readme and https://learn.microsoft.com/en-us/nuget/reference/nuspec#license. 

That ofc wouldn't handle the 3rd party notices and other files, but something to consider. 

cc @JonDouglas @aortiz-msft as this is a new experience you might want to take a look at.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs [hunk 1]
```diff
@@ -85,6 +90,18 @@ static bool IsPlaceholderFile(string path)
...
                 return false;
             }
+
+            static bool IsDocumentationFile(string path)
+            {
+                // Content files are not considered documentation. They are displayed via different means.
+                if (path.StartsWith("contentFiles/", StringComparison.OrdinalIgnoreCase))
+                    return false;
+
+                return path.StartsWith("doc/", StringComparison.OrdinalIgnoreCase)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs]: doc, documentation folders not really established patterns, and imo, we should not encourage them.  The idea is that if we encourage and unmonitored folder, we can't really monitor it without breaking customers. 

IMO, it's enough to look at .txt, .md.

**Review by nkolev92 — COMMENTED** [review]
Thanks for this contribution @drewnoakes! 
This looks like a good change we should be able to merge it soon. 

A couple of things I want to clarify. I have tagged @aortiz-msft and @JonDouglas for their input.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs (hunk 1 above)
Comment (drewnoakes) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs]: Makes sense. I can update that. I searched my disk for `project.assets.json` files containing `.txt` and `.md` and found a bunch of packages that had such folders.

Comment (drewnoakes) [reply src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs]: Fixed in 65325da213ed8668a27085f4e6a84e177e46e1a6.

**Comment by erdembayar** [comment]
Thank you for your contribution @drewnoakes 
Some tests are failed to compile, please check.
```
C:\Test\NuGet.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\SolutionExplorer\Models\Asset
sFileDependenciesSnapshotTests.cs(62,63): error CS7036: There is no argument given that corresponds to the required par
ameter 'lockFileTarget' of 'AssetsFileDependenciesSnapshot.ParseLibraries(LockFile, LockFileTarget)' [C:\Test\NuGe
t.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\NuGet.VisualStudio.Implementation.Test.csproj:
:TargetFramework=net472]
C:\Test\NuGet.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\SolutionExplorer\Models\Asset
sFileDependenciesSnapshotTests.cs(89,99): error CS7036: There is no argument given that corresponds to the required par
ameter 'lockFileTarget' of 'AssetsFileDependenciesSnapshot.ParseLibraries(LockFile, LockFileTarget)' [C:\Test\NuGe
t.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\NuGet.VisualStudio.Implementation.Test.csproj:
:TargetFramework=net472]
C:\Test\NuGet.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\SolutionExplorer\Models\Asset
sFileDependenciesSnapshotTests.cs(127,112): error CS7036: There is no argument given that corresponds to the required p
arameter 'lockFileTarget' of 'AssetsFileDependenciesSnapshot.ParseLibraries(LockFile, LockFileTarget)' [C:\Test\Nu
Get.Client\test\NuGet.Clients.Tests\NuGet.VisualStudio.Implementation.Test\NuGet.VisualStudio.Implementation.Test.cspro
j::TargetFramework=net472]
```

**Comment by drewnoakes** [comment]
> Some tests are failed to compile, please check

Fixed.

**Comment by erdembayar** [comment]
> > Some tests are failed to compile, please check
> 
> Fixed.

There was some error, I just requeued the CI build.

**Comment by erdembayar** [comment]
@drewnoakes 
Team triage: Could you be able to add unit test to prevent from regression? Please check https://github.com/NuGet/NuGet.Client/blob/dev/docs/project-overview.md for test project structure. Thanks.

**Comment by drewnoakes** [comment]
> add unit test to prevent from regression?

As mentioned in the PR description, these files are mostly declarative relationships between entities. There's no current test infrastructure for these, nor is there much to test here. The same explanation was given for me previous PRs in this area. For example, #4390.

File: src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs [hunk 2]
```diff
@@ -85,6 +90,16 @@ static bool IsPlaceholderFile(string path)
 
                 return false;
             }
+
+            static bool IsDocumentationFile(string path)
```
Comment (nkolev92) [inline src/NuGet.Clients/NuGet.VisualStudio.Implementation/SolutionExplorer/Models/AssetsFileTargetLibrary.cs:94]: This is probably one of the independently testable methods.

**Review by nkolev92 — APPROVED** [review]

**Pull request merged**